proposals/webidl-bindings/idl.md

```diff
@@ -0,0 +1,739 @@
+concepts and the concrete surface notation; we distinguish between the two by
+using a `=>` to denote an abstract production and `::=` to denote a production
+in the concrete syntax. In addition, we use the convention that non-terminals of the concrete grammar are prefixed by `W-`, as in:
+
+```
+W-Tuple ::= '(' W-Term .. W-Term ')'
+```
+
+
+#### S-Expression Notation
+
+This specification uses a variant of so-called S-Expression notation to
+represent human readable expressions. This variant -- called W-Expressions --
+has four different bracketing operators -- `()`, `[]`, `{}` and `<>`. These
+bracketing operators are distinct from one another but are otherwise equivalent.
+
+Similarly, the 'splat' operator -- written as a prefix `^` -- whose role is to unpack a W-Expression into a sequence of elements. For example, 
+
+```
+(1 2 3 4) == (1 ^(2 3) 4)
+```
+
+even though,
+
+```
+(1 2 3 4) != (1 (2 3) 4)
+```
+
+>Note: The reason for having four different bracketing operators is to support
+>readability without requiring complex parsing.
+
+## Wi-IDL Schema {#schema}
+
+The wi-IDL schema language is a language for describing types.
+
+A type is a term that denotes a collection of values. Specifically, there exists
+a unique meta-function mapping values to types (all values have a unique type).
+
+The wi-IDL type schema defines the values that a wi-IDL signature can reference.
+
+There are several different kinds of type:
+
+```
+Type => FunctionType
+Type => QuantifiedType
+Type => TypeInterface
+Type => TupleType
+Type => VariantType
+Type => NominalType
+```
+
+### Function Type {#function-type}
+
+A function type denotes the types of functions. 
+
+```
+FunctionType => TupleType '=>' Type
+```
+
+In the concrete W-Expression form, function types are written:
+
+```
+W-FunctionType ::= '(' 'func' W-TupleType W-Type ')'
+```
+
+For example, the type of a function that returns a string from a pair of
+integers has type -- as an W-Expression term:
+
+```
+(func (integer integer) string)
+```
+
+A function that returns more than one value is indicated by its return type
+being a _TupleType_.
+
+>Discussion Point: webAssembly is described in terms of a stack machine. In that
+>context, all functions simply return a tuple of values. However, in order to
+>support interoperability with systems that are not inherently stack machines,
+>this specification assumes all functions return a single value -- which may be
+>a tuple. In particular, so-called void functions are assumed to return the
+>empty tuple.
+
+A function that may raise an exception can declare it with a throws clause in
+the type:
+
+```
+FunctionType => TupleType '=>' Type 'throws' Type
+```
+or in W-Expression form:
+
+```
+W-FunctionType ::= '(' 'func' W-TupleType W-Type W-Type ')'
+```
+
+The type of an exception is not distinguished from other types. I.e., there is
+no privileged type that is used for exceptions.
```

Comment:
How does that align with Wasm's actual exception proposal?

There is a reason that exception need a uniform type, which is non-local propagation. The only way a free type can be sound is if you force translation at every level, which is not how exceptions usually operate.